Invoke all Enso benchmarks via JMH

Akirathan opened the pull request:
Closes #7323 

## Pull Request Description

### Motivation

There are a lot of Enso-only benchmarks in `test/Benchmarks` project. These benchmarks are not run on the CI (they are, but only in dry-run). We want to run these benchmarks and collect the results, such that we will see them in [engine-benchmark-results](https://enso-org.github.io/engine-benchmark-results/) along with all the other Engine bench results. These benchmarks use our Enso benchmarking infrastrcture, which basically consist of just one method - `Bench.measure`.

In order to do that, we have two choices:
- Try to invoke everything via JMH, so that JMH will provide all the data processing and data collection.
- Implement result collector to our benchmarking infrastructure.

Let's invoke every benchmark via JMH. Not only it collects the results for us, but it also does a lot of additional work like JVM preparation, forking, warmup, etc. We don't want to lose that ability.

Ideal properties of the solution:
- We don't want to manually rewrite all the benchmark sources.
  - Ideally, there is no need to any modification to any of these sources.
- We still want to be able to invoke the benchmarks as standalone scripts, just as we can do right now.

### How to use it
This PR adds two new SBT projects - `bench-processor` in `libs/scala/bench-processor` directory, and `std-benchmarks` in `std-bits/benchmarks` directory. `std-benchmarks` has just one Java class with the `main` method and with `@GenerateBenchSources` annotation.

To run one benchmark: 
```
std-benchmarks/benchOnly <bench-name-regex>
```

To force the annotation processor to rediscover new benchmarks and to regenerate the JMH sources:
```
std-benchmarks/clean; std-benchmarks/Bench/clean; std-benchmarks/Bench/compile
```

To provide additional cmdline arguments that are supported by JMH runner:
```
std-benchmarks/Bench/run -h
```

More info in `docs/infrastructure/benchmarks.md`


### Important Notes

#### The Plot

- there used to be two kinds of benchmarks: in Java and in Enso
- those in Java got quite a good treatment
- there even are results updated daily: https://enso-org.github.io/engine-benchmark-results/
- the benchmarks written in Enso used to be 2nd class citizen

#### The Revelation
This PR has the potential to fix it all!
- It designs new [Bench API](https://github.com/enso-org/enso/pull/7101/commits/88fd6fb9888d939c2d9745a7397f6e87dec4fa4d) ready for non-batch execution
- It allows for _single benchmark in a dedicated JVM_ execution
- It provides a simple way to wrap such an Enso benchmark as a Java benchmark
- thus the results of Enso and Java benchmarks are [now unified](https://github.com/enso-org/enso/pull/7101#discussion_r1257504440)

Long live _single benchmarking infrastructure for Java and Enso_!



### Checklist

Please ensure that the following checklist has been satisfied before submitting the PR:

- [x] The documentation has been updated, if necessary.
- [x] Screenshots/screencasts have been attached, if there are any visual changes. For interactive or animated visual changes, a screencast is preferred.
- [x] All code follows the
      [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md),
      [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md),
      and
      [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md)
      style guides. In case you are using a language not listed above, follow the [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md) style guide.
- All code has been tested:
  - [x] Unit tests have been written where possible.
  - [x] If GUI codebase was changed, the GUI was tested when built using `./run ide build`.

Commit:
[WIP] Invoke Enso bench via JMH

Akirathan commented on engine/runtime/src/bench/resources/org.enso.interpreter.bench.benchmarks.meso/Vector_Operations.enso:
So far, this is just a copy of [test/Benchmarks/src/Vector/Operations.enso](https://github.com/enso-org/enso/blob/develop/test/Benchmarks/src/Vector/Operations.enso). In the future, we might want to keep around some kind of config file where all the sources that should be benchmarked on the CI will be listed.

hubertp commented:
@Akirathan could we also combine this with https://github.com/enso-org/enso/issues/6554?

Akirathan commented:
> @Akirathan could we also combine this with #6554?

Well, this kind of supersedes #6554 - if this is integrated, there will be no more need for Enso benchmarks `test/Benchmarks` to provide any sophisticated output because both Engine benchmarks and Enso benchmarks will be run via JMH.

JaroslavTulach commented on engine/runtime/src/bench/scala/org/enso/interpreter/bench/IRUtils.scala:
```diff
@@ -0,0 +1,46 @@
+
+object IRUtils {
+  private def collectBenchCalls(ir: IR.Module): List[BenchCall] = {
+    val methods = ir.bindings.collect { case m : IR.Module.Scope.Definition.Method => m }
```
Btw. I will need the JMH benchmarking for #5067...

I don't like mangling with `IR` much. We should change the `Bench.measure` API to be more JMH friendly instead.

**Update**: I've just created a _builder pattern API_ for collecting all the benchmarks. Please review [this commit](https://github.com/enso-org/enso/pull/7101/commits/88fd6fb9888d939c2d9745a7397f6e87dec4fa4d) to see just my changes.

Akirathan replied:
Yes, I agree that IR manipulation is too clumsy. Let's roll out with your *builder pattern* suggestion. It should not be difficult to collect the benchmark specification in the Java code that invokes JMH with that.

JaroslavTulach committed:
Abstract from direct call to Bench.measure

JaroslavTulach committed:
Builder pattern to collect all benchmarks first and only then execute them

JaroslavTulach commented on test/Benchmarks/src/Vector/Operations.enso:
```diff
@@ -16,31 +16,40 @@ num_iterations = 10
+    all = Bench.build builder->
+        builder.group "Vector Operations" options group_builder->
+            bench group_builder
+
```
@Akirathan, my idea for integration with JMH is to change all the benchmarks like I did in [this commit](https://github.com/enso-org/enso/pull/7101/commits/88fd6fb9888d939c2d9745a7397f6e87dec4fa4d). The JMH then needs to get access to `all` and _not call_ `run_main` on them, but iterate over the benchmarks to find the one with right name and execute it in JMH.

If we agree that changing the benchmark registration to the builder pattern is a way to go, then you can prepare a generic JMH benchmark that would cooperate with `Bench`, `Bench_Options` & co.

JaroslavTulach commented on distribution/lib/Standard/Test/0.0.0-dev/src/Bench.enso:
```diff
@@ -1,7 +1,71 @@
 type Bench
+    All (groups : Vector Bench)
+    Group (name:Text) (configuration:Bench_Options) (specs : Vector Bench)
+    Spec (name:Text) (code : Any -> Any)
```
`Bench` is the API which holds all the found `Group` and `Spec` instances. This API can be used in a simple way via `run_main` or _disassembled to pieces_ by JMH...

JaroslavTulach commented on distribution/lib/Standard/Test/0.0.0-dev/src/Bench.enso:
```diff
@@ -1,7 +1,71 @@
+    Spec (name:Text) (code : Any -> Any)
+
+    build : (Bench_Builder -> Any) -> Bench
+    build fn =
```
I was experimenting with various ways to construct the benchmark cases, but I have found nothing better than old good builder pattern. I tried various tricks with `State`, but at the end `Vector.new_builder` (with its mutability) seems to work the best and without major surprises.

```haskell
from Standard.Base import all
from Standard.Test import Bench

all = Bench.build builder->
    builder.group "Group1" (Bench.options.iter 3 . size 1) group->
        group.specify "bench1"
            1 * 1
        group.specify "bench2"
            1 * 2
    builder.group "group2" (Bench.options.iter 5 . size 1) group->
        group.specify "another1"
            2 * 1
        group.specify "another2"
            2 * 2

main = all.run_main
```
if this is found acceptable, then we should change our `Test_Suite` harness to use the same pattern. For integration with VSCode it is important to collect tests first (even without executing them), present them to the user and allow user to choose one of them to be executed.

Akirathan replied:
I was thinking about a similar pattern. This suggestion looks good. The only thing that remains to be solved is how to implement `Bench.run_main` such that when run from CLI, it will behave in the old way, and when run from JMH, it will be a no-op. I believe that this can be tweaked in the JMH. Generally, we just need to think about how to achieve that once this script is invoked from CLI, it will immediately run all the benchmarks and when invoked from JMH, it will gather the descriptions.

TL;DR; I believe the builder pattern, similar to the one we are already using for our Enso tests, is the only way to go.

JaroslavTulach replied:
>  when run from CLI, it will behave in the old way

Why it should behave the old way? Is there any benefit in _not collecting_ all the cases first? I can only think of more memory needed to hold all the cases initialized, but even that can be solved by `@Before` (or `@Setup`) and `@After` annotations/principle.

I certainly need the tests to behave in such a way - _collect all the cases first_ and only then let the VSCode invoke (for example) one of them.

radeusgd replied:
I really like collecting the cases, as you say - we can then easily do a 'dry run' and allow more interactive test selection. Moreover, knowing the amount of tests beforehand gives us first steps towards adding some progress indicators (% of tests already completed, maybe ETA) - I love progress bars so I'd love to see this (ofc this is not a very important feature but a cool nice to have one).

For the test suite there's one thing to consider - our tests tend to do some 'preparation' before running the particular test cases (`Test.specify` branches) - some of it inside of `Test.group` and even some outside of it. That preparation tends to be side-effecting (e.g. 
 establishing DB connection, creating some base tables in a DB) and takes some time (creating a table over a connection tends to be in the magnitude of 300ms, not 3ms IIRC).

When gathering just descriptions - we probably don't want to run this preparatory stuff as it would only be a waste of time. But we cannot easily 'fold' it into `Test.specify` because that would force the preparation to be re-run for every single test case - I'm not sure what will be the cost of that - we may have to measure it.

I think ideally we'd need to refactor the tests in such a way that the structure can be gathered without running these preparatory actions. It may require some work to achieve that.

JaroslavTulach replied:
Running the _"preparation"_ only when needed is an essential part of testing API. That's why I'd like the libraries team to design it, at the end. I can think of the _"JUnit style"_ - e.g. adding 
```ruby
type Group_Builder
    before : Function -> ???
    after: Function -> ???
```
each group would call all registered `before` handlers before running `specify` and then it would call `after`. However how to access the results of `before` in `specify` (in a side-effect free language) remains a mystery to me!?

However we may not need any special construct. Enough to specify lazy value. E.g. define
```ruby
type My_Data
    Connection ~sql
```
and then create an instance in `group` scope and access the `instance.sql` only when needed. It will initialize just once, at first warmup iteration.

How to handle cleanup isn't completely clear to me. Could `Managed_Resource` help somehow? Let's hope it would and let's thus leave the topic of _before&after_ for later. It is not essential to running of a _single benchmark_ that this PR tackles.

radeusgd replied:
I imagine we could have `specify` take an additional argument which will be filled in by `before`?

i.e.

```
builder.before <|
    h1 = ...
    h2 = ...
    My_Helpers.Instance h1 h2

builder.specify "foo" helpers->
    helpers.h1.foo_bar helpers.h2
```
what do you think?
It's not the most pretty IMHO but it could work ok.

Akirathan replied:
Yet another suggestion is to have `before` as part of `Bench_Options` and specify it as function to function.

```diff
diff --git a/distribution/lib/Standard/Test/0.0.0-dev/src/Bench.enso b/distribution/lib/Standard/Test/0.0.0-dev/src/Bench.enso
--- a/distribution/lib/Standard/Test/0.0.0-dev/src/Bench.enso	(revision 74ac8d759e54c9880477f06eb8907d83b8107580)
+++ b/distribution/lib/Standard/Test/0.0.0-dev/src/Bench.enso	(date 1689700961515)
@@ -4,13 +4,16 @@
 
 type Bench_Options
     ## PRIVATE
-    Impl iter_size num_iters
+    Impl iter_size num_iters before_func
 
     size : Integer -> Bench_Options
-    size self v = Bench_Options.Impl v self.num_iters
+    size self v = Bench_Options.Impl v self.num_iters self.before_func
 
     iter : Integer -> Bench_Options
-    iter self v = Bench_Options.Impl self.iter_size v
+    iter self v = Bench_Options.Impl self.iter_size v self.before_func
+
+    before : (Nothing -> Any) -> Bench_Options
+    before self fn = Bench_Options.Impl self.iter_size self.num_iters fn
 
     to_text self = "[iter_size=" + self.iter_size.to_text + ", num_iters=" + self.num_iters.to_text + "]"
 
@@ -45,7 +48,7 @@
         Bench.All b.to_vector
 
     options : Bench_Options
-    options = Bench_Options.Impl -1 -1
+    options = Bench_Options.Impl -1 -1 Nothing
 
     fold : Any -> (Any -> Bench -> Bench -> Any) -> Any
     fold self value fn = case self of
@@ -60,7 +63,9 @@
         self.fold Nothing _-> g-> s->
             c = g.configuration
             IO.println <| "Benchmarking  " + s.name + " configuration: " + c.to_text
-            Bench.measure (s.code 0) s.name c.iter_size c.num_iters
+            data = c.before_func Nothing
+            IO.println <| "Before data = " + data.to_text
+            Bench.measure (s.code data) s.name c.iter_size c.num_iters
             IO.println <| "Benchmarking of " + s.name + " finished"
 
     ## Measure the amount of time it takes to execute a given computation.

```

And usage:
```
from Standard.Base import all
import Standard.Test.Bench.Bench
import Standard.Test.Bench.Bench_Options

before_func =
    # Do some side effects?
    # ...
    IO.println "[mylog] In before_func"
    [1,2,3,4,5]

tmp = Bench.options . size 10
options = Bench.options . size 10 . iter 5 . before (_ -> before_func)

all = Bench.build builder->
    builder.group "group_1" options group_builder->
        group_builder.specify "bench_1" <|
            vec -> vec + [10]

main =
    all.run_main

```

Akirathan replied:
Yet another suggestion is to have `before` as part of `Bench_Options` and specify it as function to function.

```diff
diff --git a/distribution/lib/Standard/Test/0.0.0-dev/src/Bench.enso b/distribution/lib/Standard/Test/0.0.0-dev/src/Bench.enso
--- a/distribution/lib/Standard/Test/0.0.0-dev/src/Bench.enso	(revision 74ac8d759e54c9880477f06eb8907d83b8107580)
+++ b/distribution/lib/Standard/Test/0.0.0-dev/src/Bench.enso	(date 1689700961515)
@@ -4,13 +4,16 @@
 
 type Bench_Options
     ## PRIVATE
-    Impl iter_size num_iters
+    Impl iter_size num_iters before_func
 
     size : Integer -> Bench_Options
-    size self v = Bench_Options.Impl v self.num_iters
+    size self v = Bench_Options.Impl v self.num_iters self.before_func
 
     iter : Integer -> Bench_Options
-    iter self v = Bench_Options.Impl self.iter_size v
+    iter self v = Bench_Options.Impl self.iter_size v self.before_func
+
+    before : (Nothing -> Any) -> Bench_Options
+    before self fn = Bench_Options.Impl self.iter_size self.num_iters fn
 
     to_text self = "[iter_size=" + self.iter_size.to_text + ", num_iters=" + self.num_iters.to_text + "]"
 
@@ -45,7 +48,7 @@
         Bench.All b.to_vector
 
     options : Bench_Options
-    options = Bench_Options.Impl -1 -1
+    options = Bench_Options.Impl -1 -1 Nothing
 
     fold : Any -> (Any -> Bench -> Bench -> Any) -> Any
     fold self value fn = case self of
@@ -60,7 +63,9 @@
         self.fold Nothing _-> g-> s->
             c = g.configuration
             IO.println <| "Benchmarking  " + s.name + " configuration: " + c.to_text
-            Bench.measure (s.code 0) s.name c.iter_size c.num_iters
+            data = c.before_func Nothing
+            IO.println <| "Before data = " + data.to_text
+            Bench.measure (s.code data) s.name c.iter_size c.num_iters
             IO.println <| "Benchmarking of " + s.name + " finished"
 
     ## Measure the amount of time it takes to execute a given computation.

```

And usage:
```
from Standard.Base import all
import Standard.Test.Bench.Bench
import Standard.Test.Bench.Bench_Options

before_func =
    # Do some side effects?
    # ...
    IO.println "[mylog] In before_func"
    [1,2,3,4,5]

tmp = Bench.options . size 10
options = Bench.options . size 10 . iter 5 . before (_ -> before_func)

all = Bench.build builder->
    builder.group "group_1" options group_builder->
        group_builder.specify "bench_1" <|
            vec -> vec + [10]

main =
    all.run_main

```

JaroslavTulach replied:
Discussing `before` is completely fine, but please leave the implementation for another PR. Let's focus on getting this PR merged during this iteration with the goal to have an `sbt` command to invoke (some of) the Enso benchmarks and get `bench-result.xml` compatible output.

Commit:
Improve type ascriptions of Bench type

Commit:
[WIP] Add LibBenchRunner

Akirathan commented:
Current idea of the solution (sketched in https://github.com/enso-org/enso/pull/7101/commits/f6dd4237d5ab3ca3011c1c4ae60358ba8f4c5b4e):
- There will be one class `LibBenchRunner` with `main` method that:
  - Collects all the benchmark specifications from `test/Benchmarks` project
  - Parses JMH specific cmdline arguments and merges them with some custom arguments
    - Inspired by `Main.java` from JMH project that parses all the arguments.
  - Delegates to a proper benchmark
- One class `LibBench` with a single method annotated with `@Benchmark`.
  - Needed in order for the JMH runner to recognize it as a benchmark.
  - This method takes parameters and extracts the benchmark name from it and delegates to an appropriate benchmark

Another alternative is to generate many Java sources with many methods annotated with `@Benchmark` based on discovered benchmarks from `test/Benchmarks`, but that is too complicated.

IR manipulation is out of the question - better to refactor `test/Benchmarks` than manipulate with IR.

This idea assumes that all the benchmarks in `test/Benchmarks` are migrated to the [builder pattern](https://github.com/enso-org/enso/pull/7101/commits/88fd6fb9888d939c2d9745a7397f6e87dec4fa4d).

JaroslavTulach commented on engine/runtime/src/bench/java/org/enso/interpreter/bench/benchmarks/meso/LibBenchRunner.java:
```diff
@@ -0,0 +1,66 @@
+import org.openjdk.jmh.runner.options.OptionsBuilder;
+
+// TODO: Move to different project?
+public class LibBenchRunner {
```
What is this? How is this supposed to be used? Ideally I want to execute:
```
sbt:runtime> benchOnly VectorOperations
```
and all the `VectorOperations` benchmarks would be executed. I don't see anything like that in `LibBenchRunner`, neithre in `LibBench`.

**Update1**: The code wasn't even compilable! Fixed in 58c1e0d

JaroslavTulach replied:
Since f5a9aae you can use:
```bash
sbt:enso> benchOnly VectorOperations
org.enso.interpreter.bench.benchmarks.meso.VectorOperations.newConstant
org.enso.interpreter.bench.benchmarks.meso.VectorOperations.newRandom
org.enso.interpreter.bench.benchmarks.meso.VectorOperations.newVector
org.enso.interpreter.bench.benchmarks.meso.VectorOperations.sum
```
in JMH. Verified by [running the benchmarks on the CI](https://github.com/enso-org/enso/actions/runs/5500703727) and getting the new results including `VectorOperations` runs:

![VectorOperations benchmarks](https://github.com/enso-org/enso/assets/26887752/f0859827-244f-4041-ac58-2f6607bec860)

jdunkerley replied:
This looks like this is now part of the engine space? The VectorOperations are now part of the engine runtime tests.

Are we retiring the old benchmark project and will we need to get every new benchmark (which we clearly should do more of) approved by engine team going forward?

JaroslavTulach committed:
Making the project compilable (again)

JaroslavTulach committed:
Executed few suites from Vector/Operations.enso in JMH

JaroslavTulach committed:
Merge remote-tracking branch 'origin/develop' into wip/akirathan/enso-bench-jmh

JaroslavTulach committed:
Bench is the type, not Group

JaroslavTulach committed:
sumStatistic benchmark

JaroslavTulach committed:
Generate double values

Commit:
Add prototype of bench suites discovery

Commit:
Move LibBenchRunner to a separate project `bench-libs`

JaroslavTulach commented on engine/runtime/src/bench/scala/org/enso/interpreter/bench/IRUtils.scala:
```diff
@@ -1,8 +1,7 @@
 
-import org.enso.compiler.core.IR
-
 object IRUtils {
```
There `IRUtils` can now be removed from this PR, I assume.

Akirathan replied:
Sure, done in 0ee38dfb8444eae3e841a578916b6148f1d611e9

JaroslavTulach commented on tools/bench-libs/src/main/java/org/enso/benchmarks/libs/LibBenchRunner.java:
```diff
@@ -60,7 +66,37 @@ private void collectBenchSpecs() {
+
+  private static Set<BenchSuite> collectAllBenchSpecs() {
+    String benchmarksProjectRootDir = "test/Benchmarks/src";
+    File singleFile = new File("/home/pavel/dev/enso/engine/runtime/src/bench/resources/org.enso.interpreter.bench.benchmarks.meso/simple.enso");
```
Let's try to locate the benchmarks via relative way. For example like https://github.com/enso-org/enso/pull/7101/files/f6dd4237d5ab3ca3011c1c4ae60358ba8f4c5b4e..9d9f328ec384e303d4cdf6a23e6deccd2cf8c755#diff-a8c733e77d2301c4075061f5f53e29e7aea47170134995974796c14ba20231c7R40

Commit:
Get rid of IRUtils.scala

JaroslavTulach commented on build.sbt:
```diff
@@ -1722,6 +1722,38 @@ lazy val `distribution-manager` = project
   .dependsOn(pkg)
   .dependsOn(`logging-utils`)
 
+lazy val `bench-libs` = (project in file("tools/bench-libs"))
```
I believe `std-bits/benchmarks` would be a better and more consistent name for a project that contains Java code and deals with Enso sources.

Akirathan replied:
Relocated to `bench-libs` project to `std-bits/benchmarks`.

Commit:
Merge branch 'develop' into wip/akirathan/enso-bench-jmh

# Conflicts:
#	distribution/lib/Standard/Test/0.0.0-dev/src/Bench.enso
#	test/Benchmarks/src/Vector/Operations.enso

Commit:
Add SpecCollector

Commit:
LibBenchRunner prints help

Commit:
Merge branch 'develop' into wip/akirathan/enso-bench-jmh

Commit:
LibBenchRunner can list benchmarks

Commit:
Fix Operations.enso to provide specs and not to run them

JaroslavTulach commented on tools/bench-libs/src/main/java/org/enso/benchmarks/libs/LibBenchRunner.java:
```diff
@@ -52,30 +45,45 @@ public static void main(String[] args) {
-    for (BenchSuite benchSpec : benchSpecs) {
-      for (BenchGroup group : benchSpec.groups()) {
-        System.out.println("Discovered group: " + group.name());
+    File ensoRootDir = Paths.get(System.getProperty("user.dir")).toFile();
```
Wouldn't it be better to use `LibBenchRunner.class.getProtectionDomain().getCodeSource().getLocation()` to obtain the location of the class itself and search from it?

Akirathan replied:
Sure. I didn't know that this was an option. Thanks for the suggestion.

JaroslavTulach commented on tools/bench-libs/src/main/java/org/enso/benchmarks/libs/SpecCollector.java:
```diff
@@ -1,56 +1,90 @@
+    if (!allSuitesVar.isNull()) {
+      try {
+        return module
+            .invokeMember(Module.EVAL_EXPRESSION, "all")
```
The name doesn't have to be `all` - I used it, but we can choose anything more descriptive `bench`, `all_benchmarks`, etc.

Akirathan replied:
This is already implemented, as I have noted in some of my previous comments.

JaroslavTulach commented on tools/bench-libs/src/main/java/org/enso/benchmarks/libs/SpecCollector.java:
```diff
@@ -1,56 +1,90 @@
+    try (Stream<Path> stream = Files.walk(rootDir.toPath())) {
+      return stream
+          .map(Path::toFile)
+          .filter(file -> file.isFile() && file.canRead() && file.getName().endsWith(".enso"))
```
We are walking all the files and calling `all` in each of it. That's possible approach.

However currently the test harness as well as benchmark harness define `Main.enso` and list all the tests and benchmarks from there. If we stick to this approach, we wouldn't need to scan all the file. We'd just executed single, fixed entry point like `Main.all_benchmarks` and got them all.

I have no preference myself, however. CCing @jdunkerley and @radeusgd.

Akirathan replied:
This behavior is already changed. No more scanning all the source files. Explained in https://github.com/enso-org/enso/pull/7101/files/8f3d761e697af6ddb5e43e63f48aa899ab426a66#r1281685901

Commit:
Add simple annotation processor

Commit:
Move bench-libs to std-bits/benchmarks

Commit:
Integrate annotation processor into bench-libs

JaroslavTulach commented on build.sbt:
```diff
@@ -1770,14 +1770,23 @@ lazy val `bench-libs` = (project in file("tools/bench-libs"))
-      "org.openjdk.jmh" % "jmh-core"                  % jmhVersion,
-      "org.openjdk.jmh" % "jmh-generator-annprocess"  % jmhVersion,
-      "org.graalvm.sdk" % "graal-sdk"                 % graalMavenPackagesVersion   % "provided",
+      "org.netbeans.api" % "org-openide-util-lookup"   % netbeansApiVersion,
```
**TCR**:Are you using only `@ServiceProvider` annotation from the [lookup library](https://bits.netbeans.org/dev/javadoc/org-openide-util-lookup/overview-summary.html)? Then it is enough to have just a `% "provided"` dependency.

Akirathan replied:
Yes, I am using only the `@ServiceProvider` annotation. I have changed the dependency to provided in df993839d0142255b8c90f8658769120c1f550af

JaroslavTulach commented on build.sbt:
```diff
@@ -1779,14 +1795,6 @@ lazy val `bench-libs` = (project in file("std-bits/benchmarks"))
-      // fixing incremental compilation setup
-      Seq(
-        "-processor",
-        "org.netbeans.modules.openide.util.ServiceProviderProcessor"
```
Áááá... so this was the reason the processor couldn't be executed! Removing `-processor` option is good. Now the processor is going to be discovered via `META-INF/services`.

I think you can now merge the two projects into one - if you move `LibBenchRunner` into `src/bench/java`...

Akirathan replied:
This comment is outdated. We now have two projects - `std-benchmarks` and `bench-processor`.

Commit:
Add truffle and runtime to annotation processor's class path

JaroslavTulach commented on build.sbt:
```diff
@@ -1806,9 +1808,11 @@ lazy val `bench-libs` = (project in file("std-bits/benchmarks"))
           .map(_.data)
           .mkString(File.pathSeparator)
+      // Only run ServiceProvider processor and ignore those defined in META-INF, thus
+      // fixing incremental compilation setup
```
This comment is copy left. I don't think we restrict annotation processors in this place anymore.

JaroslavTulach commented on project/FrgaalJavaCompiler.scala:
```diff
@@ -9,19 +9,20 @@
  */
 
-import sbt._
+import sbt.*
```
Looks like an unnecessary _no-semantic change_. Needless to say I prefer non-start - e.g. explicit imports.

**TCA**: prefer explicit imports.

Akirathan replied:
Reverted in 7df60c5346f11f7baaeb55727243bd1290519065 . Turns out my IDE thinks that Scala sources in the `project` directory are Scala 3 rather than Scala 2.

JaroslavTulach commented on project/FrgaalJavaCompiler.scala:
```diff
@@ -78,6 +86,15 @@ object FrgaalJavaCompiler {
+    var noLimitModules = false
+    val strippedJArgs: ArrayBuffer[String] = ArrayBuffer()
+    for (strippedJArg <- jArgs.map(_.stripPrefix("-J"))) {
+      if (strippedJArg == "--no-limit-modules") {
```
**TCR**: Remove support for `--no-limit-modules` - looks like it is enough to included `org.graalvm.locator` in the list of enabled modules to let Graal SDK find all languages.

Akirathan replied:
I have tried removing `--no-limit-modules` in the most recent version of the PR - 7df60c5346f11f7baaeb55727243bd1290519065, but it does not work. Even if I add `org.graalvm.locator` to the list of `--limit-modules`, the `BenchProcessor` fails on `NoClassDefFoundError` in `org.enso.pkg.COnfig#fromYaml` with stack trace:
```
"main@1" prio=5 tid=0x1 nid=NA runnable
  java.lang.Thread.State: RUNNABLE
	  at cats.syntax.EitherObjectOps$.catchNonFatal$extension(either.scala:390)
	  at io.circe.yaml.Parser.parseSingle(Parser.scala:39)
	  at io.circe.yaml.Parser.parse(Parser.scala:29)
	  at io.circe.yaml.parser.package$.parse(package.scala:18)
	  at org.enso.pkg.Config$.fromYaml(Config.scala:256)
	  at org.enso.pkg.PackageManager.$anonfun$loadPackage$2(Package.scala:317)
	  at org.enso.pkg.PackageManager$$Lambda$288/0x00000007c05073f8.apply(Unknown Source:-1)
	  at scala.util.Using$.$anonfun$apply$1(Using.scala:262)
	  at scala.util.Using$.apply(Using.scala:113)
	  at org.enso.pkg.PackageManager.readConfig$1(Package.scala:317)
	  at org.enso.pkg.PackageManager.loadPackage(Package.scala:322)
	  at org.enso.interpreter.runtime.EnsoContext.lambda$initialize$2(EnsoContext.java:146)
	  at org.enso.interpreter.runtime.EnsoContext$$Lambda$286/0x00000007c0506648.apply(Unknown Source:-1)
	  at java.util.Optional.map(Optional.java:260)
	  at org.enso.interpreter.runtime.EnsoContext.initialize(EnsoContext.java:143)
	  at org.enso.interpreter.EnsoLanguage.initializeContext(EnsoLanguage.java:156)
	  at org.enso.interpreter.EnsoLanguage.initializeContext(EnsoLanguage.java:62)
	  at com.oracle.truffle.api.TruffleLanguage$Env.postInit(TruffleLanguage.java:3588)
	  at com.oracle.truffle.api.LanguageAccessor$LanguageImpl.postInitEnv(LanguageAccessor.java:287)
	  at com.oracle.truffle.polyglot.PolyglotLanguageContext.ensureInitialized(PolyglotLanguageContext.java:719)
	  at com.oracle.truffle.polyglot.PolyglotContextImpl.getBindings(PolyglotContextImpl.java:1217)
	  at com.oracle.truffle.polyglot.PolyglotContextDispatch.getBindings(PolyglotContextDispatch.java:98)
	  at org.graalvm.polyglot.Context.getBindings(Context.java:564)
	  at org.enso.benchmarks.processor.BenchProcessor.getModule(BenchProcessor.java:137)
	  at org.enso.benchmarks.processor.BenchProcessor.process(BenchProcessor.java:117)
	  at com.sun.tools.javac.processing.JavacProcessingEnvironment.callProcessor(JavacProcessingEnvironment.java:1024)
	  at com.sun.tools.javac.processing.JavacProcessingEnvironment.discoverAndRunProcs(JavacProcessingEnvironment.java:940)
	  at com.sun.tools.javac.processing.JavacProcessingEnvironment.access$2300(JavacProcessingEnvironment.java:112)
	  at com.sun.tools.javac.processing.JavacProcessingEnvironment$Round.run(JavacProcessingEnvironment.java:1268)
	  at com.sun.tools.javac.processing.JavacProcessingEnvironment.doProcessing(JavacProcessingEnvironment.java:1383)
	  at com.sun.tools.javac.main.JavaCompiler.processAnnotations(JavaCompiler.java:1246)
	  at com.sun.tools.javac.main.JavaCompiler.compile(JavaCompiler.java:925)
	  at com.sun.tools.javac.main.Main.compile(Main.java:317)
	  at com.sun.tools.javac.main.Main.compile(Main.java:176)
	  at com.sun.tools.javac.Main.compile2(Main.java:76)
	  at jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(NativeMethodAccessorImpl.java:-1)
	  at jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
	  at jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	  at java.lang.reflect.Method.invoke(Method.java:568)
	  at org.frgaal.Main.lambda$compile$0(Main.java:24)
	  at org.frgaal.Main$$Lambda$24/0x00000007c00c0900.run(Unknown Source:-1)
	  at org.frgaal.AbstractMain.doRun(AbstractMain.java:27)
	  at org.frgaal.Main.compile(Main.java:21)
	  at org.frgaal.Main.main(Main.java:17)
```

For now, I am leaving the `--no-limit-modules` there, with the exception that there will be a documented constant in `FrgaalJavaCompiler.scala` instead of hardcoded "-J--no-limit-modules".

Akirathan replied:
Refactored to a documented constant in `FrgaalJavaCompiler.scala` in 6676b8cde22e31eb569c5bd0bdffa63eff6165cf

JaroslavTulach reviewed (changes requested):
I am afraid the `--no-limit-modules` isn't the _Real McCoy_. Rather I'd like to know what JDK modules are missing to run the annotation processor properly?

JaroslavTulach commented on project/FrgaalJavaCompiler.scala:
```diff
@@ -213,9 +230,15 @@ object FrgaalJavaCompiler {
+      } else {
+        Seq(
+          "--limit-modules",
+          "java.base,jdk.zipfs,jdk.internal.vm.compiler.management",
```
I think I got the processor working with simple:
```java
        Seq(
          "--limit-modules",
          "java.base,jdk.zipfs,jdk.internal.vm.compiler.management,org.graalvm.locator",
        )
```
e.g. looks like adding `org.graalvm.locator` is the change that makes the SDK work.

Akirathan replied:
This does not work anymore. See my previous comment https://github.com/enso-org/enso/pull/7101#discussion_r1281601789.

Commit:
BenchProcessor creates JMH sources for all benchmarks

Commit:
Put generated code into src_managed directory

Commit:
Remove benchmark discovery code from LibBenchRunner

JaroslavTulach commented on std-bits/bench-processor/src/main/java/org/enso/benchmarks/processor/BenchProcessor.java:
```diff
@@ -0,0 +1,233 @@
+    javaSrcFileWriter.append("      ).build();\n");
+    javaSrcFileWriter.append("    \n");
+    javaSrcFileWriter.append("    Value module = ctx.getBindings(LanguageInfo.ID).invokeMember(MethodNames.TopScope.GET_MODULE, \"" + moduleName + "\");\n");
+    javaSrcFileWriter.append("    File benchProjectDir = new File(\"" + benchRootDir.getAbsolutePath() + "\");\n");
```
**TCR**: The generated source code contains fully qualified name:
```java
    File benchProjectDir = new File("/home/enso/enso/test/Benchmarks");
    File languageHomeOverride = new File("/home/enso/enso/distribution/component");
```
that means the generated code is different on each computer. It'd be better if the generated code was always the same.

Please replace with relative lookup with the help of `ProtectionDomain.getCodeSource().getLocation()`.

Akirathan replied:
Done in f5622840c9ec7e3a4f978a57898f21b238f88945

JaroslavTulach commented on std-bits/bench-processor/src/main/java/org/enso/benchmarks/processor/GenerateBenchSources.java:
```diff
@@ -0,0 +1,15 @@
+ */
+@Retention(RetentionPolicy.CLASS)
+@Target(ElementType.TYPE)
+public @interface GenerateBenchSources {
```
**TCA**: I'd add two arguments to this annotation:
- location of `.enso` file to evaluate
- name of a symbol to invoke to obtain all `BenchSuite`

The `@Target` could also be changed to be only or allow also a `PACKAGE`. Then people could create their own `package-info.java` and:
```java
@GenerateBenchSources(
  file="test/Benchmarks/src/Main.enso",
  invoke="all_benchmarks"
)
package my.benchmarks;
```
in their own project and get their own set of benchmarks.

Akirathan replied:
In the most recent version of this PR - https://github.com/enso-org/enso/blob/f5622840c9ec7e3a4f978a57898f21b238f88945/lib/scala/bench-processor/src/main/java/org/enso/benchmarks/processor/GenerateBenchSources.java, there are three arguments:
- `projectRootPath` that is a relative path to the project root
- `moduleName` that is FQN of the module
- `variableName` - symbol name that is used for the lookup of the benchmark specification.

I believe this is a good first set of the parameters for the annotation. We could probably replace `moduleName` with `fileName`, or remove it altogether, but now, this feels like a detail.

JaroslavTulach commented on engine/runtime/src/bench/java/org/enso/interpreter/bench/benchmarks/meso/VectorOperations.java:
**TCR**: Can be deleted.

Akirathan replied:
Done in 882ae8ab3df0f421ffc19157ead60c13b7609ffa

JaroslavTulach commented on engine/runtime/src/bench/resources/org.enso.interpreter.bench.benchmarks.meso/Vector_Operations.enso:
**TCR**: No longer needed. Delete.

Akirathan replied:
Done in 882ae8ab3df0f421ffc19157ead60c13b7609ffa

JaroslavTulach commented on engine/runtime/src/bench/resources/org.enso.interpreter.bench.benchmarks.meso/simple.enso:
**TCR**:  Not needed anymore. Delete.

Akirathan replied:
Done in 882ae8ab3df0f421ffc19157ead60c13b7609ffa

JaroslavTulach commented on project/FrgaalJavaCompiler.scala:
```diff
@@ -9,19 +9,20 @@
 import sbt.util.Logger
-import xsbti.{PathBasedFile, Reporter, VirtualFile, Logger => XLogger}
-import xsbti.compile.{IncToolOptions, Output, JavaCompiler => XJavaCompiler}
+import xsbti.{PathBasedFile, Reporter, VirtualFile, Logger as XLogger}
```
The usage of `as` breaks `scalafmtSbt`. Returning `=>` back fixes the formatting again.

Akirathan replied:
Fixed. Turns out my IDE thinks that this source file should be in Scala 3.

Commit:
Do not use generate package name for generated sources

Commit:
Incorporate bench and benchOnly commands in bench-libs project

Commit:
Move LibBenchRunner under src/bench from src/main

Commit:
Benchmark/run/javaOptions gets fullClasspath

Commit:
Generate source into Compile/sourceManaged.

So that they can be tracked by an IDE.

Commit:
Use fully qualified module names in BenchProcessor

Commit:
Add ability to attach debugger to the annotation processors.

Commit:
Add the benchmarks documentation

Commit:
Discard spurious warnings from javac for generated sources

Commit:
Remove some temporary logging

Commit:
Fail annotation processing properly instead of throwing exceptions

Commit:
Fix SourceVersion warning

Commit:
Catch exception, not throwable

Commit:
cosmetics

Commit:
Merge branch 'develop' into wip/akirathan/enso-bench-jmh

# Conflicts:
#	engine/runtime/src/main/scala/org/enso/compiler/PackageRepository.scala

JaroslavTulach commented on std-bits/bench-processor/src/main/java/org/enso/benchmarks/BenchSuiteWrapper.java:
```diff
@@ -0,0 +1,52 @@
+import java.util.Optional;
+import org.graalvm.polyglot.Value;
+
+public class BenchSuiteWrapper {
```
`final`

JaroslavTulach commented on docs/infrastructure/benchmarks.md:
```diff
@@ -0,0 +1,78 @@
+# Benchmarks
+In this document, we describe the benchmark types used for the runtime - micro benchmarks in
+the section [Engine JMH microbenchmarks](#engine-jmh-microbenchmarks) and meso benchmarks
```
What is a _"meso"_? I have seen the term being used few times, but I have no idea why it is called _"meso"_?

Akirathan replied:
I thought that the term *mesobenchmark* is wide spread in the software engineering community. **microbenchmarks** track performance metrics on the lowest level, for example, a specific operation on `Vector` type. **macrobenchmarks** track performance of a real-life-like application, for example, [SPECjbb2015](https://www.spec.org/jbb2015/). Finally, **mesobenchmarks** are something in between of macro and micro. 

Nevertheless, the usage of *mesobenchmark* terminology is unnecessary in this case, so I have removed it in 62f34801a925290558a5593816bb130d1ace0743

JaroslavTulach replied:
I tried to google _"meso benchmark"_ out before writing my complain. Maybe I am not properly profiled, but I get no usable results. For [example](https://medium.com/the-wind-vane/newa-meso-micro-challenge-phase-2-complex-terrain-98efeb03a23a)..

JaroslavTulach commented on docs/infrastructure/benchmarks.md:
```diff
@@ -0,0 +1,78 @@
+- [micro benchmarks](#engine-jmh-microbenchmarks) located directly in the `runtime` SBT project.
+  These benchmarks are written in Java, and are used to measure the performance of
+  specific parts of the engine.
+- [standard library benchmarks](#standard-library-benchmarks) (meso benchmarks) located in the `test/Benchmarks` Enso project.
```
Again: Why _"meso"_? Because writing them _"je maso"_?

JaroslavTulach committed:
Easier way to debug the runtime/benchOnly command

JaroslavTulach commented on docs/infrastructure/benchmarks.md:
```diff
@@ -0,0 +1,78 @@
+modify the parameter to the `@Measurement` annotation.
+
+### Debugging the benchmarks
+Currently, the best way to debug the benchmark is to set the `@Fork` annotation to 0, and to
```
Got an idea: 95e6deaeba270cfc0ad70df89b5423d871dd0dba

Akirathan replied:
yes, that is a better idea.

JaroslavTulach commented on docs/infrastructure/benchmarks.md:
```diff
@@ -0,0 +1,78 @@
+
+## Standard library benchmarks
+Unlike the micro benchmarks, these benchmarks are written entirelly in Enso and located in the
+`test/Benchmarks` Enso project. Sometimes, we call those *meso* benchmarks.
```
> Sometimes, we call those *meso* benchmarks.

**TCR**: Do we? Why? Please justify or remove the term usage.

JaroslavTulach commented on docs/infrastructure/benchmarks.md:
```diff
@@ -0,0 +1,78 @@
+For more sophisticated approach, run the benchmarks via JMH launcher.
+
+### Running via JMH launcher
+The JMH launcher is located in `std-bits/benchmarks` directory, as `bench-libs` SBT project.
```
All the other `std-bits` projects are named `std-xyz`. 
**TCR**:Change the name of the benchmarking project to `std-benchmarks` ...

**TCA**: ... or (better) disperse benchmarks into existing `std-xyz` projects. `Vector_Operations` would be in `std-base` and some table benchmarks would go to `std-table`.

Akirathan replied:
Rename to `std-benchmarks` in ea5315f2abc1fa87dcbbe07b2ae6885459e40708.

> TCA: ... or (better) disperse benchmarks into existing std-xyz projects

Let's keep this option open until we get the initial feedbacks of the usage. Let's keep it simple for now and let's have just a separate `std-benchmarks` project for now. I will keep this conversation open.

JaroslavTulach commented on docs/infrastructure/benchmarks.md:
```diff
@@ -0,0 +1,78 @@
+To access the full flexibility of the JMH launcher, run it via `Bench/run` - for example,
+to see the help message: `Bench/run -h`.
+
+To debug a single benchmark, set the `javaOptions` with something like `set javaOptions += "-agentlib:jdwp=transport=dt_socket,server=y,suspend=y,address=localhost:8000"`,
```
**TCR**: Please integrate with `withDebug --debugger` just like in the case of `runtime` project. Use the `-Dbench.compileOnly=true` property check to prevent forking just like 95e6deaeba270 does.

Akirathan replied:
Done in ed1909dc9d975611e199bdaa99cbc4283edd33c5

JaroslavTulach commented on docs/infrastructure/benchmarks.md:
```diff
@@ -0,0 +1,78 @@
+that can be invoked manually on a specific branch as well.
+The results are kept in the artifacts produced from the actions.
+In `tools/performance/engine-benchmarks` directory, there is a simple Python
+script for collecting and processing the results. See the README in that directory
```
```suggestion
script for collecting and processing the results. See the [README in that directory](../../tools/performance/engine-benchmarks/README.md)
```

JaroslavTulach commented on docs/infrastructure/benchmarks.md:
```diff
@@ -0,0 +1,78 @@
+This script is invoked regularly on a private machine and the results are
+published in [https://enso-org.github.io/engine-benchmark-results/](https://enso-org.github.io/engine-benchmark-results/).
+
+This is not meant as a permanent solution.
```
**TCA**: Remove the sentence. Wishes and opinions don't belong to technical documentation.

JaroslavTulach commented on docs/CONTRIBUTING.md:
```diff
@@ -415,6 +416,32 @@ sbt:project-manager> withDebug run --debugger
 
 read more about [debugging Java & Enso code](debugger/README.md).
 
+#### Debugging annotation processors
```
**TCR**: Line 417 says:

> read more about [debugging Java & Enso code](debugger/README.md).

shouldn't this paragraph be in the `debugger/README.md` file? It is a bit too specific to deserve a place in general contributing document.

Akirathan replied:
> shouldn't this paragraph be in the debugger/README.md file?

Probably yes, I have rewritten the documentation about engine debugging in 00e6cf39804f8a0e9a00fbb102ef65f3b7524701 - I hope for the better.

JaroslavTulach commented on project/FrgaalJavaCompiler.scala:
```diff
@@ -9,25 +9,27 @@
 
   val frgaal      = "org.frgaal" % "compiler" % "19.0.1" % "provided"
   val sourceLevel = "19"
+  val debugArg = "-J-agentlib:jdwp=transport=dt_socket,server=y,suspend=y,address=localhost:8000"
```
**TCR**: Please use `WithDebugCommand.DEBUG_OPTION` instead to be across all the invocations of debugger by `sbt`.

Akirathan replied:
Can't be simply done, as explained in https://github.com/enso-org/enso/pull/7101#discussion_r1282095635

JaroslavTulach commented on docs/CONTRIBUTING.md:
```diff
@@ -415,6 +416,32 @@ sbt:project-manager> withDebug run --debugger
+  - Delete all the compiled class files along with all the generated sources by
+    the annotation processor. This ensures that the annotation processor will
+    be invoked.
+- `set javacOptions += FrgaalJavaCompiler.debugArg`
```
**TCR**: Please use `WithDebugCommand.DEBUG_OPTION` instead to be consistent with all `sbt` debugger invocations.

Akirathan replied:
While I like consistency as much as you like it, I don't think this is a good idea to use `WithDebugCommand.DEBUG_OPTION` for `javacOptions`. The whole `FrgaalJavaCompiler` machinery first invokes `java ... -jar frgaal-compiler.jar` with redirected output, and after that process is finished, it prints (and potentially filters) the output from the compilation process to stdout. With `WithDebugCommand.DEBUG_OPTION`, the JVM tries to attach to an already listening debugger right after startup, whereas `FrgaalJavaCompiler.debugArg` makes JVM suspend and wait for the debugger to attach. I can intercept this argument and print a helper message before the JVM process is started, informing the user that there is a JVM process waiting for the debugger.

Besides, I have tried to do it via `WithDebugCommand.DEBUG_OPTION` and I am unable to make it work. Let's stick with this "special" `FrgaalJavaCompiler.debugArg` for now. There is no need to unify debugging of unit tests and debugging of annotation processors anyway.

JaroslavTulach commented on std-bits/bench-processor/src/main/java/org/enso/benchmarks/processor/SpecCollector.java:
```diff
@@ -0,0 +1,110 @@
+        .build();
+  }
+
+  public Collection<BenchSuiteWrapper> collectAllBenchSpecs() {
```
**TCA**: I don't understand why `SpecCollector` and `BenchSuiteWrapper` are in different packages? I'd even merge them into a single source file.

Akirathan replied:
`SpecCollector` is in `org.enso.benchmarks.processor` because it is coupled with the annotation processing and code generation process, whereas `BenchSuiteWrapper` (that is now renamed to `ModuleBenchSuite`) is in `org.enso.benchmarks` package because it is a *universal* type meant to be used from the generated code.

JaroslavTulach commented on build.sbt:
```diff
@@ -1769,6 +1769,103 @@ lazy val `distribution-manager` = project
   .dependsOn(pkg)
   .dependsOn(`logging-utils`)
 
+lazy val `bench-processor` = (project in file("std-bits/bench-processor"))
```
**TCA**: Do @jdunkerley or @radeusgd like this processor among `std-bits` modules? I am not 100% comfortable with the location. I'd rather hide it under `lib` directory even it is a very special library that builds on top of whole Enso runtime...

Akirathan replied:
I don't like the location either. I suggest putting it to `lib/scala/bench-processor`.

JaroslavTulach commented on std-bits/bench-processor/src/main/java/org/enso/benchmarks/BenchGroup.java:
```diff
@@ -0,0 +1,9 @@
+import java.util.List;
+
+public interface BenchGroup {
+  String name();
```
**TCA**: Hide these files as inner classes of `SpecCollector` or `BenchSuiteWrapper`, etc.

Akirathan replied:
I tried to hide those, but failed to do so. These interfaces are used by the generated sources that are in a different project, in a different package. Therefore, for now, I am leaving these classes as public.

JaroslavTulach commented on std-bits/bench-processor/src/main/java/org/enso/benchmarks/BenchSuiteWrapper.java:
```diff
@@ -0,0 +1,52 @@
+  private final BenchSuite suite;
+  private final String moduleQualifiedName;
+
+  public BenchSuiteWrapper(BenchSuite suite, String moduleQualifiedName) {
```
The name `BenchSuiteWrapper` is a bit scary. If this is the main API for people to use, then:

**TCA**: Hide current `BenchSuite` & co. interfaces more and give `BenchSuiteWrapper` nicer name. Plus make sure the class is collocated with `SpecColector` - they work in orchestration.

Akirathan replied:
You are right, it is not a good name. I have done some refactoring in 2e57598fba196f5f6f27a9d77055e25c8f0ffd67, hopefully for the better.

Akirathan replied:
Note that hiding `BenchSuite` and other interfaces is not entirely possible, since they are used from the JMH generated code, which might be in a different project.

JaroslavTulach commented on std-bits/bench-processor/src/main/java/org/enso/benchmarks/BenchSuiteWrapper.java:
```diff
@@ -0,0 +1,52 @@
+    return moduleQualifiedName;
+  }
+
+  public Value getDefaultInputArgument() {
```
I don't like this _forward thinkingness_. There doesn't seem to be any reason for this method to exist right now, so we shouldn't have it.

**TCA**: Remove unnecessary methods from `BenchSuiteWrapper`.

Akirathan replied:
Done in 80564b00fae2d7e046bb17713033dbd8c83133f9

JaroslavTulach commented on test/Benchmarks/src/Vector/Operations.enso:
```diff
@@ -49,13 +49,10 @@ collect_benches group_builder =
-    all = Bench.build builder->
-        builder.group "Vector Operations" options group_builder->
-            collect_benches group_builder
+all = Bench.build builder->
```
**TCA**: Rename to `all_benchmarks`. Use `@GenerateBenchSources(invoke="all_benchmarks")` to connect the Java and Enso code together.

JaroslavTulach commented on std-bits/benchmarks/src/bench/java/org/enso/benchmarks/libs/LibBenchRunner.java:
```diff
@@ -0,0 +1,54 @@
+import org.openjdk.jmh.runner.options.CommandLineOptions;
+
+@GenerateBenchSources
+public class LibBenchRunner {
```
The name `LibBenchRunner` is misleading - it doesn't do anything with Enso! 

**TCR**: Move the `main` class/method to the `bench-processor` project and reuse it.
**TCA**: Shall not [BenchmarksRunner](https://github.com/enso-org/enso/blob/7a934cadccd5ebec3f418835c59f040839068146/engine/runtime/src/bench/java/org/enso/interpreter/bench/BenchmarksRunner.java#L18) just depend on `bench-processor` project? Shall not the XML read/write functionality be in that `bench-processor` project?
**TCA**: I'd like `bench-processor` project to be moved away from `std-bits` to `lib/java/benchmarks`

JaroslavTulach commented on std-bits/benchmarks/src/bench/java/org/enso/benchmarks/libs/LibBenchRunner.java:
```diff
@@ -0,0 +1,54 @@
+
+    // TODO[pm]: Process the results in the same way as we do in `runtime/bench`.
+    System.out.println("Results:");
+    for (RunResult result : results) {
```
**TCR**: Share code with https://github.com/enso-org/enso/blob/7a934cadccd5ebec3f418835c59f040839068146/engine/runtime/src/bench/java/org/enso/interpreter/bench/BenchmarksRunner.java#L18

Akirathan replied:
Done. This code is now in `lib/scala/bench-processor` project in `org.enso.benchmarks.runner.BenchRunner#run`.

JaroslavTulach reviewed (changes requested):
I have just gone thru the code changes again and realized there is a lot of comments of mine. I decided to classify them:
- **TCR** - _technical change required_ - something I am convinced have to be addressed or argued about
- **TCA** - _technical change advised_ - consider the change, but no need to argue
- the rest - are just my opinions

I can't wait to see how my _libraries friends_ react when they see and use these changes for real.

Commit:
Restrict dependencies scopes

Commit:
Merge branch 'origin/wip/akirathan/enso-bench-jmh' of github.com:enso-org/enso into wip/akirathan/enso-bench-jmh

Commit:
Restructure debugging docs, reformat.

Commit:
Remove meso benchmark terminology

Commit:
Remove unecessary method from BenchSUiteWrapper

Commit:
Some refactoring.

Commit:
Move bench-processor from std-bits to lib/scala/bench-processor

Commit:
Add Benchmark runner code to bench-processor

Commit:
Delete unused benchmark code

Commit:
Distinct.enso benchmark uses the new bench builder API

Commit:
Main module of test/Benchmarks collects all bench specs

Commit:
Fix jakarta.xml dependency

Commit:
Add module-discovery parameters to GenerateBenchSources annotation

JaroslavTulach commented on docs/debugger/README.md:
```diff
@@ -8,96 +8,14 @@ order: 0
+This folder contains all documentation pertaining to the debugging facilities
+used by Enso, broken up as follows:
+
+- [**The Enso Debugger Protocol:**](./protocol.md) The protocol for the REPL
```
Separating the documentation into individual files seems like a good idea.

Commit:
Add tests for SpecCollector

Commit:
BenchProcessor supports new module-discovery parameters

Commit:
Some improvements of bench-libs SBT project.

- No explicit mainClass setting needed.
- Remove specified excludeMsg log manager

Commit:
Unify report collecting

Commit:
Add Utils class

Commit:
Add some tests

Commit:
No star import in Scala sources in project

Commit:
--no-limit-modules is not a hardcoded, but a documentted constant in `FrgaalJavaCompiler`.

Commit:
Filter out spurious warning messages from javac

Commit:
Reformat bench-libs project in sbt - group settings by the configuration.

Commit:
Generated source does not contain absolute paths

Commit:
Rename `bench-libs` sbt project to `std-benchmarks`.

Commit:
In std-benchmarks project, withDebug command works the same as in runtime benchmarks

Commit:
Add some docs

JaroslavTulach commented on project/FrgaalJavaCompiler.scala:
```diff
@@ -211,11 +235,18 @@ object FrgaalJavaCompiler {
-      val forkArgs = (jArgs ++ Seq(
-        "--limit-modules",
-        "java.base,jdk.zipfs,jdk.internal.vm.compiler.management",
+      val limitModulesArgs = if (noLimitModules) {
```
**TCR**: If this patch eliminates the need for `--no-limit-modules`, then apply it:
```diff
diff --git build.sbt build.sbt
index 53d7cb01d4..191b058df9 100644
--- build.sbt
+++ build.sbt
@@ -1842,7 +1842,6 @@ lazy val `std-benchmarks` = (project in file("std-bits/benchmarks"))
           .map(_.data)
           .mkString(File.pathSeparator)
       Seq(
-        FrgaalJavaCompiler.noLimitModulesArg,
         s"-J-Dtruffle.class.path.append=$appendClasspath"
       )
     },
diff --git project/FrgaalJavaCompiler.scala project/FrgaalJavaCompiler.scala
index 5a4c3a59df..870e1683bb 100644
--- project/FrgaalJavaCompiler.scala
+++ project/FrgaalJavaCompiler.scala
@@ -29,16 +29,6 @@ object FrgaalJavaCompiler {
   val frgaal      = "org.frgaal" % "compiler" % "19.0.1" % "provided"
   val sourceLevel = "19"
 
-  /** By default, the Frgaal compiler is invoked with `--limit-modules` argument that limits
-    * the set of modules that are accessible at compile time to just a small list of internal
-    * JDK modules. Pass this flag to the compiler in order to disable this behavior.
-    * This is usable in cases when an annotation processor invoked by Frgaal needs to access
-    * classes from other modules.
-    *
-    * Note that to pass this flag to the compiler, set `javacOptions` in the sbt shell with
-    * `set javacOptions += FrgaalJavaCompiler.noLimitModulesArg`.
-    */
-  val noLimitModulesArg = "-J--no-limit-modules"
   val debugArg =
     "-J-agentlib:jdwp=transport=dt_socket,server=y,suspend=y,address=localhost:8000"
 
@@ -79,11 +69,6 @@ object FrgaalJavaCompiler {
   }
 
   /** Helper method to launch programs.
-    *
-    * If -J--no-limit-modules is passed to javac, no --limit-modules option is passed to frgaal.
-    * With -J--no-limit-modules, one can use any class path in annotation processors.
-    * All the options with "-J" prefix are prepended to the frgaal command line, instead of passing
-    * them via an argument file. Moreover, the "-J" prefix is stripped.
     */
   def launch(
     javaHome: Option[Path],
@@ -98,9 +83,7 @@ object FrgaalJavaCompiler {
   ): Boolean = {
     val (jArgs, nonJArgs)     = options.partition(_.startsWith("-J"))
     val debugAnotProcessorOpt = jArgs.contains(debugArg)
-    val noLimitModules        = jArgs.contains(noLimitModulesArg)
     val strippedJArgs = jArgs
-      .filterNot(_ == noLimitModulesArg)
       .map(_.stripPrefix("-J"))
     val outputOption = CompilerArguments.outputOption(output)
     val sources = sources0 map { case x: PathBasedFile =>
@@ -235,16 +218,10 @@ object FrgaalJavaCompiler {
     val allArguments = outputOption ++ frgaalOptions ++ nonJArgs ++ sources
 
     withArgumentFile(allArguments) { argsFile =>
-      val limitModulesArgs = if (noLimitModules) {
-        Seq()
-      } else {
-        // Need to disable standard compiler tools that come with used jdk and replace them
-        // with the ones provided with Frgaal.
-        Seq(
+      val limitModulesArgs =Seq(
           "--limit-modules",
-          "java.base,jdk.zipfs,jdk.internal.vm.compiler.management,org.graalvm.locator"
+          "java.base,jdk.zipfs,jdk.internal.vm.compiler.management,org.graalvm.locator,java.desktop,java.net.http"
         )
-      }
       // strippedJArgs needs to be passed via cmd line, and not via the argument file
       val forkArgs = (strippedJArgs ++ limitModulesArgs ++ Seq(
         "-jar",
```
I am able to run `sbt:std-benchmarks> benchOnly Random_Integer_Vector_Distinct` and `std-bits/benchmarks/bench-report.xml` file is properly generated.

Akirathan replied:
Yes, that works for me too. Thank you for the patch, applied in 2664ded4c0a56ad8ecca43a342b0c58c3b3e3a60

Commit:
Remove --no-limit-modules

Commit:
fmt docs

JaroslavTulach reviewed (approved):
I've just used this support to [execute Greg's benchmarks](https://github.com/enso-org/enso/pull/7460/files#r1283283587) and see the results in IGV. I think the support is good enough.

radeusgd commented on docs/CONTRIBUTING.md:
```diff
@@ -756,7 +757,8 @@ interface of the runner prints all server options when you execute it with
 - `--server`: Runs the Language Server
-- `--root-id <uuid>`: Content root id.
+- `--root-id <uuid>`: Content root id. The Language Server chooses one randomly,
+  so can pass any valid UUID.
```
nitpick:
```suggestion
  so any valid UUID can be passed.
```

or
```suggestion
  so you can pass any valid UUID.
```

radeusgd commented on docs/debugger/chrome-devtools.md:
```diff
@@ -0,0 +1,41 @@
+
+## Tips and tricks
+
+- Use `env JAVA_OPTS=-Dpolyglot.inspect.Path=enso_debug` to set the chrome to
+  use a fixed URL. In this case the URL is
+  `devtools://devtools/bundled/js_app.html?ws=127.0.0.1:9229/enso_debug`
```
Wow I did not know that, so cool!

radeusgd commented on docs/infrastructure/benchmarks.md:
```diff
@@ -0,0 +1,102 @@
+
+## Standard library benchmarks
+
+Unlike the Engine micro benchmarks, these benchmarks are written entirelly in
```
```suggestion
Unlike the Engine micro benchmarks, these benchmarks are written entirely in
```

radeusgd commented on docs/infrastructure/benchmarks.md:
```diff
@@ -0,0 +1,102 @@
+
+The `std-benchmarks` SBT project supports `bench` and `benchOnly` commands, that
+work the same as in the `runtime` project, with the exception that the benchmark
+name does not have to be specified as a fully qualified name, but as a regular
```
Does this not work in `runtime`?

radeusgd replied:
(using regexp to filter benchmarks)

Akirathan replied:
Not yet, I will look into that in one of the follow-up PRs and unify the functionality even more.

radeusgd commented on docs/infrastructure/benchmarks.md:
```diff
@@ -0,0 +1,102 @@
+`Bench/run -w 3 -i 2 New_Vector`.
+
+Whenever you add or delete any benchmarks from `test/Benchmarks` project, the
+generated JMH sources need to be recompiled with `Bench/clean; Bench/compile`.
```
Is it not possible to do this without `clean`? Would it be possible for the compiler to detect that it needs to be rebuilt based on source changes or is that too problematic?

Akirathan replied:
It could theoretically be possible. But definitelly too complicated. Since it is not clear which Enso sources should be tracked.

radeusgd commented on lib/scala/bench-processor/src/main/java/org/enso/benchmarks/BenchConfig.java:
```diff
@@ -0,0 +1,5 @@
+package org.enso.benchmarks;
+public interface BenchConfig {
+  int size();
```
What does `size` mean here?

Akirathan replied:
Added docs in 495e6e66c7d1aceb066dd58d64133f46dc938e7f

radeusgd commented on lib/scala/bench-processor/src/main/java/org/enso/benchmarks/BenchSpec.java:
```diff
@@ -0,0 +1,12 @@
+ */
+public interface BenchSpec {
+  String name();
+  Value code();
```
Is it really the `code` or more like an executable _thunk_?

Akirathan replied:
It's exactly the same as `code` field in `Bench.Spec`, as mentioned in the javadoc. It's just a polyglot `org.graalvm.polyglot.Value` that can be executed. I guess that technically it is a thunk, but what difference does it make here?

radeusgd commented on project/FrgaalJavaCompiler.scala:
```diff
@@ -224,6 +232,14 @@ object FrgaalJavaCompiler {
       val cwd         = new File(new File(".").getAbsolutePath).getCanonicalFile
       val javacLogger = new JavacLogger(log, reporter, cwd)
       var exitCode    = -1
+      if (debugAnotProcessorOpt) {
+        log.info(
+          s"Frgaal compiler is about to be launched with $debugArg, which means that" +
+          " it will wait for a debugger to attach. The output from the compiler is by default" +
+          " redirected, therefore \"Listening to the debugger\" message will not be displayed." +
+          " You should attach the debugger now."
+        )
+      }
```
Wow, thanks for adding this message!

Akirathan replied:
Without that message, you would have no idea that the `java` process is waiting for a debugger, because the output is redirected.

radeusgd commented on test/Benchmarks/src/Vector/Distinct.enso:
```diff
@@ -1,31 +1,27 @@
+options = Bench.options . size 100 . iter 20
 
-## Bench Utilities ============================================================
+random_vec = Utils.make_random_vec 100000
+uniform_vec = Vector.fill 100000 1
+random_text_vec = random_vec.map .to_text
+uniform_text_vec = random_vec.map .to_text
```
This is not good.

Enso has a subtle difference between top-level and scoped definitions, that plays a very crucial role here.

A scoped definition is computed (unless it's a block...) and stored. A top-level definition is re-computed on every access.

Essentially you have turned values _computed once at initialization of the suite_ into 0-argument methods that are _recomputed on each access_.

This changes the semantics of these benchmarks.

We used to be just computing the time needed to compute the `distinct` operation. Now we are computing the total time it takes to both generate the vector _and_ compute distinct.

I don't think that's right.

I know that we don't have the 'setup' pattern yet. For now I'd just run this inside of the `group`. It's not perfect and will slow down gathering benchmarks, so maybe for future we need a better solution. 

On the other hand, I'm not 100% convinced that changing semantics is bad here. I don't think it is good but feel free to convince me otherwise.

Akirathan replied:
> A scoped definition is computed (unless it's a block...) and stored. A top-level definition is re-computed on every access.

I have not realized that. That might be problematic here - we don't want a different `random_vec` in each iteration. I will try to revert that. My goal here was just to demonstrate the usage of the builder pattern that is syntactically a bit nicer than the usage in `Operations.enso`.

Akirathan replied:
On second thought, let me take care of this in follow-up PRs, I have now added it to a task list in #7489.

Commit:
Add some docs

Commit:
Minor fixes in the docs

Commit:
Merge branch 'develop' into wip/akirathan/enso-bench-jmh

Akirathan commented on engine/runtime/src/bench/java/org/enso/interpreter/bench/BenchmarksRunner.java:
```diff
@@ -39,7 +41,8 @@ public BenchmarkItem run(String label) throws RunnerException, JAXBException {
         .measurementIterations(1)
-        .warmupIterations(0);
+        .warmupIterations(0)
+        .forks(0);
```
Setting forks to 0 when `bench.CompileOnly = true` seems to have a negative side effect when all the benchmarks are run on the CI - it seems to cause StackOverflow errors, like [this one](https://github.com/enso-org/enso/actions/runs/5762870936/job/15623476613#step:10:1624)

Akirathan replied:
I have reverted 11e0a12c6fc449afede4a325659e071daef0104c as a workaround for now. In the upcoming PRs, I will try to unify and improve the debugging experience of the benchmarks. Tracking this in #7489.

Commit:
Revert "Easier way to debug the runtime/benchOnly command"

This reverts commit 95e6deaeba270cfc0ad70df89b5423d871dd0dba.

Commit:
fmt docs

JaroslavTulach commented on project/FrgaalJavaCompiler.scala:
```diff
@@ -29,6 +29,9 @@ object FrgaalJavaCompiler {
   val sourceLevel = "19"
 
+  val debugArg =
+    "-J-agentlib:jdwp=transport=dt_socket,server=y,suspend=y,address=localhost:8000"
```
I believe this option shall be unified with `WithDebugCommand` setup... I wanted to mention that before, but maybe I haven't done so...

Akirathan replied:
You have mentioned that already. And the answer is buried somewhere within dozens of comments in this PR. The brief answer is that we need the `javac` process (more specifically, in this case, `java -jar frgaal-compiler.jar` process) to **wait** for the debugger and not to start without it. That is because the `java -jar frgaal-compiler.jar` process's output is piped, so the compiler process can finish even before the user notices that they could attach a debugger.

Moreover, using `WithDebugCommand.DEBUG_ARG` requires some more refactoring - for that, we would need to remove some files in `project` directory from `org.enso.build` package.

TL;DR; it would be unnecessarily complicated.
